File: is_permutation.cpp
#include <iostream>
#include <string>
#include <map>
#include <cassert>

using namespace std;

bool isPermutationHT(string str1,string str2){

	if(str1.length() != str2.length()){
		return false;
	}

	if (str1.compare(str2) == 0){
		return false;
	}

	map<char, int> ht;
	char* cstr1= (char*) str1.c_str();
	for(int i = 0; i<str1.length(); i++) {

		if(ht[cstr1[i]] == 0){
			ht[cstr1[i]] = ht[cstr1[i]] + 1;
		}
		else {
			ht[cstr1[i]] = ht[cstr1[i]] + 1;
		}
	}

	char* cstr2 = (char*) str2.c_str();
	for(int i=0; i<str2.length(); i++){

		if(ht[cstr2[i]] != 0){
			ht[cstr2[i]] = ht[cstr2[i]] - 1;

		 if(ht[cstr2[i]] < 0){
		 	return false;
		 }
		}

		else {
			return false;
		}

		return true;
	}
}

int main(int argc, char **argv){

	string string1 = "sumeyya";
	string string2 = "suyyaem";

	cout << isPermutationHT(string1,string2) << endl;

	return 0;
}

File: is_permutation.py

def is_permutation(string1, string2):
	if len(string1) != len(string2):
		return False
	if string1 == string2:
		return False
	ht = dict()
	for char in string1:
		ht[char] = ht[char] + 1 if char in ht else 1
	for char in string2:
		if char in ht:
			ht[char] = ht[char] - 1
			if ht[char] < 0:
				return False
		else:
			return False
	return True

def is_permutation_sorted(string1, string2):
	if len(string1) != len(string2):
		return False
	if string1 == string2:
		return False
	str1 = sorted(string1)
	str2 = sorted(string2)
	if str1 == str2:
		return True
	return False


def main():
	string1 = "omeeerr "
	string2 = "eeerrmo "
	ip = is_permutation(string1, string2)
	print(ip)

if __name__ == "__main__":
	main()
